common/dpcpp/slab_hash.hpp

```diff
@@ -23,49 +24,118 @@ template <size_t A, size_t B, size_t P> struct DefaultHasher {
 };
 
 template <typename T> struct SlabNode {
-  SlabNode() = default;
   SlabNode(T el) {
     for (int i = 0; i < SLAB_SIZE; i++) {
       data[i] = el;
     }
   }
 
   T data[SLAB_SIZE];
-  sycl::global_ptr<SlabNode<T>> next = nullptr;
+  sycl::device_ptr<SlabNode<T>> next = nullptr;
 };
 
 template <typename T> struct SlabList {
   SlabList() = default;
 
-  sycl::global_ptr<SlabNode<T>> root;
+  sycl::device_ptr<SlabNode<T>> root;
+};
+
+template <typename T> struct HeapMaster {
```

Comment:
I assume this is an implementation detail, if so, consider hiding it in some kind of a `detail` namespace.